File: frontend/src/api/trade.ts
import type { PageData } from '../types/api'
import type { ProductSummary } from '../types/catalog'
import type {
  Address,
  AddressPayload,
  AdminOrderDetail,
  AdminOrderSummary,
  AdminReview,
  CartSummary,
  OrderDetail,
  OrderSummary,
  PaymentResult,
  Review,
  OrderStatus,
  Wallet,
  WalletTransaction,
} from '../types/trade'
import { http } from './http'

export async function getFavorites(page = 1, pageSize = 20) {
  return (
    await http.get('/favorites', { params: { page, page_size: pageSize } })
  ).data.data as PageData<ProductSummary>
}

export async function getFavoriteStatus(productId: number) {
  return (await http.get(`/favorites/${productId}/status`)).data.data as boolean
}

export async function addFavorite(productId: number) {
  return (await http.post(`/favorites/${productId}`)).data.data as ProductSummary
}

export async function removeFavorite(productId: number) {
  await http.delete(`/favorites/${productId}`)
}

export async function getAddresses() {
  return (await http.get('/addresses')).data.data as Address[]
}

export async function createAddress(payload: AddressPayload) {
  return (await http.post('/addresses', payload)).data.data as Address
}

export async function updateAddress(id: number, payload: Partial<AddressPayload>) {
  return (await http.patch(`/addresses/${id}`, payload)).data.data as Address
}

export async function deleteAddress(id: number) {
  await http.delete(`/addresses/${id}`)
}

export async function getCart() {
  return (await http.get('/cart')).data.data as CartSummary
}

export async function addCartItem(skuId: number, quantity: number) {
  return (await http.post('/cart/items', { sku_id: skuId, quantity })).data.data as CartSummary
}

export async function updateCartItem(
  itemId: number,
  payload: { quantity?: number; selected?: boolean },
) {
  return (await http.patch(`/cart/items/${itemId}`, payload)).data.data as CartSummary
}

export async function selectAllCartItems(selected: boolean) {
  return (await http.put('/cart/selection', { selected })).data.data as CartSummary
}

export async function deleteCartItem(itemId: number) {
  return (await http.delete(`/cart/items/${itemId}`)).data.data as CartSummary
}

export async function getWallet() {
  return (await http.get('/wallet')).data.data as Wallet
}

export async function rechargeWallet(amount: number) {
  return (await http.post('/wallet/recharge', { amount })).data.data as Wallet
}

export async function getWalletTransactions(page = 1, pageSize = 20) {
  return (
    await http.get('/wallet/transactions', { params: { page, page_size: pageSize } })
  ).data.data as PageData<WalletTransaction>
}

export async function createOrder(addressId: number, buyerRemark?: string) {
  return (
    await http.post('/orders', { address_id: addressId, buyer_remark: buyerRemark || null })
  ).data.data as OrderDetail
}

export async function getOrders(page = 1, pageSize = 20) {
  return (await http.get('/orders', { params: { page, page_size: pageSize } })).data
    .data as PageData<OrderSummary>
}

export async function getOrder(id: number) {
  return (await http.get(`/orders/${id}`)).data.data as OrderDetail
}

export async function payOrder(id: number) {
  return (await http.post(`/orders/${id}/pay`)).data.data as PaymentResult
}

export async function cancelOrder(id: number) {
  return (await http.post(`/orders/${id}/cancel`)).data.data as OrderDetail
}

export async function completeOrder(id: number) {
  return (await http.post(`/orders/${id}/complete`)).data.data as OrderDetail
}

export async function createReview(payload: {
  order_item_id: number
  rating: number
  content: string
  anonymous: boolean
  image_urls?: string[]
}) {
  return (await http.post('/reviews', payload)).data.data as Review
}

export async function getProductReviews(productId: number, page = 1, pageSize = 20) {
  return (
    await http.get(`/catalog/products/${productId}/reviews`, {
      params: { page, page_size: pageSize },
    })
  ).data.data as PageData<Review>
}

export async function getAdminOrders(
  page = 1,
  pageSize = 20,
  orderStatus?: OrderStatus,
) {
  return (
    await http.get('/admin/orders', {
      params: { page, page_size: pageSize, order_status: orderStatus },
    })
  ).data.data as PageData<AdminOrderSummary>
}

export async function getAdminOrder(id: number) {
  return (await http.get(`/admin/orders/${id}`)).data.data as AdminOrderDetail
}

export async function shipAdminOrder(id: number) {
  return (await http.post(`/admin/orders/${id}/ship`)).data.data as AdminOrderDetail
}

export async function completeAdminOrder(id: number) {
  return (await http.post(`/admin/orders/${id}/complete`)).data.data as AdminOrderDetail
}

export async function getAdminReviews(page = 1, pageSize = 20) {
  return (
    await http.get('/admin/reviews', { params: { page, page_size: pageSize } })
  ).data.data as PageData<AdminReview>
}

export async function updateReviewVisibility(id: number, visible: boolean) {
  return (await http.patch(`/admin/reviews/${id}`, { visible })).data.data as AdminReview
}

File: backend/app/api/v1/routes/trade.py
from typing import Annotated

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from backend.app.api.dependencies import get_current_user
from backend.app.core.config import get_settings
from backend.app.core.database import get_db
from backend.app.core.exceptions import AuthorizationError
from backend.app.core.responses import ApiResponse
from backend.app.models.user import User
from backend.app.schemas.catalog import ProductSummary
from backend.app.schemas.common import PageData
from backend.app.schemas.trade import (
    AddressCreate,
    AddressPublic,
    AddressUpdate,
    CartAddRequest,
    CartSelectRequest,
    CartSummary,
    CartUpdateRequest,
    CheckoutRequest,
    OrderDetail,
    OrderSummary,
    PaymentResult,
    RechargeRequest,
    ReviewCreate,
    ReviewPublic,
    WalletPublic,
    WalletTransactionPublic,
)
from backend.app.services.trade import TradeService

router = APIRouter(tags=["交易闭环"])
DbSession = Annotated[AsyncSession, Depends(get_db)]
CurrentUser = Annotated[User, Depends(get_current_user)]


@router.get("/favorites", response_model=ApiResponse[PageData[ProductSummary]])
async def list_favorites(
    session: DbSession,
    user: CurrentUser,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=50)] = 20,
) -> ApiResponse[PageData[ProductSummary]]:
    return ApiResponse(
        data=await TradeService(session).list_favorites(
            user.id, page=page, page_size=page_size
        )
    )


@router.get("/favorites/{product_id}/status", response_model=ApiResponse[bool])
async def favorite_status(
    product_id: int, session: DbSession, user: CurrentUser
) -> ApiResponse[bool]:
    return ApiResponse(data=await TradeService(session).is_favorite(user.id, product_id))


@router.post("/favorites/{product_id}", response_model=ApiResponse[ProductSummary])
async def add_favorite(
    product_id: int, session: DbSession, user: CurrentUser
) -> ApiResponse[ProductSummary]:
    return ApiResponse(
        message="已加入收藏",
        data=await TradeService(session).add_favorite(user.id, product_id),
    )


@router.delete("/favorites/{product_id}", response_model=ApiResponse[None])
async def remove_favorite(
    product_id: int, session: DbSession, user: CurrentUser
) -> ApiResponse[None]:
    await TradeService(session).remove_favorite(user.id, product_id)
    return ApiResponse(message="已取消收藏")


@router.get("/addresses", response_model=ApiResponse[list[AddressPublic]])
async def list_addresses(session: DbSession, user: CurrentUser) -> ApiResponse[list[AddressPublic]]:
    return ApiResponse(data=await TradeService(session).list_addresses(user.id))


@router.post(
    "/addresses",
    response_model=ApiResponse[AddressPublic],
    status_code=status.HTTP_201_CREATED,
)
async def create_address(
    payload: AddressCreate, session: DbSession, user: CurrentUser
) -> ApiResponse[AddressPublic]:
    return ApiResponse(
        message="收货地址已保存",
        data=await TradeService(session).create_address(user.id, payload),
    )


@router.patch("/addresses/{address_id}", response_model=ApiResponse[AddressPublic])
async def update_address(
    address_id: int,
    payload: AddressUpdate,
    session: DbSession,
    user: CurrentUser,
) -> ApiResponse[AddressPublic]:
    return ApiResponse(
        data=await TradeService(session).update_address(user.id, address_id, payload)
    )


@router.delete("/addresses/{address_id}", response_model=ApiResponse[None])
async def delete_address(
    address_id: int, session: DbSession, user: CurrentUser
) -> ApiResponse[None]:
    await TradeService(session).delete_address(user.id, address_id)
    return ApiResponse(message="收货地址已删除")


@router.get("/cart", response_model=ApiResponse[CartSummary])
async def get_cart(session: DbSession, user: CurrentUser) -> ApiResponse[CartSummary]:
    return ApiResponse(data=await TradeService(session).get_cart(user.id))


@router.post(
    "/cart/items", response_model=ApiResponse[CartSummary], status_code=status.HTTP_201_CREATED
)
async def add_to_cart(
    payload: CartAddRequest, session: DbSession, user: CurrentUser
) -> ApiResponse[CartSummary]:
    return ApiResponse(
        message="已加入购物车",
        data=await TradeService(session).add_to_cart(user.id, payload),
    )


@router.patch("/cart/items/{item_id}", response_model=ApiResponse[CartSummary])
async def update_cart_item(
    item_id: int,
    payload: CartUpdateRequest,
    session: DbSession,
    user: CurrentUser,
) -> ApiResponse[CartSummary]:
    return ApiResponse(data=await TradeService(session).update_cart_item(user.id, item_id, payload))


@router.put("/cart/selection", response_model=ApiResponse[CartSummary])
async def select_all_cart_items(
    payload: CartSelectRequest, session: DbSession, user: CurrentUser
) -> ApiResponse[CartSummary]:
    return ApiResponse(
        data=await TradeService(session).select_all_cart_items(user.id, payload.selected)
    )


@router.delete("/cart/items/{item_id}", response_model=ApiResponse[CartSummary])
async def delete_cart_item(
    item_id: int, session: DbSession, user: CurrentUser
) -> ApiResponse[CartSummary]:
    return ApiResponse(data=await TradeService(session).delete_cart_item(user.id, item_id))


@router.get("/wallet", response_model=ApiResponse[WalletPublic])
async def get_wallet(session: DbSession, user: CurrentUser) -> ApiResponse[WalletPublic]:
    return ApiResponse(data=await TradeService(session).get_wallet(user.id))


@router.post("/wallet/recharge", response_model=ApiResponse[WalletPublic])
async def recharge_wallet(
    payload: RechargeRequest, session: DbSession, user: CurrentUser
) -> ApiResponse[WalletPublic]:
    if not get_settings().enable_demo_recharge:
        raise AuthorizationError("演示充值功能未启用")
    return ApiResponse(
        message="充值成功",
        data=await TradeService(session).recharge_wallet(user.id, payload.amount),
    )


@router.get(
    "/wallet/transactions",
    response_model=ApiResponse[PageData[WalletTransactionPublic]],
)
async def list_wallet_transactions(
    session: DbSession,
    user: CurrentUser,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=50)] = 20,
) -> ApiResponse[PageData[WalletTransactionPublic]]:
    return ApiResponse(
        data=await TradeService(session).list_wallet_transactions(
            user.id, page=page, page_size=page_size
        )
    )


@router.post(
    "/orders",
    response_model=ApiResponse[OrderDetail],
    status_code=status.HTTP_201_CREATED,
)
async def create_order(
    payload: CheckoutRequest, session: DbSession, user: CurrentUser
) -> ApiResponse[OrderDetail]:
    return ApiResponse(
        message="订单已创建，请完成支付",
        data=await TradeService(session).create_order(user.id, payload),
    )


@router.get("/orders", response_model=ApiResponse[PageData[OrderSummary]])
async def list_orders(
    session: DbSession,
    user: CurrentUser,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=50)] = 20,
) -> ApiResponse[PageData[OrderSummary]]:
    return ApiResponse(
        data=await TradeService(session).list_orders(user.id, page=page, page_size=page_size)
    )


@router.get("/orders/{order_id}", response_model=ApiResponse[OrderDetail])
async def get_order(
    order_id: int, session: DbSession, user: CurrentUser
) -> ApiResponse[OrderDetail]:
    return ApiResponse(data=await TradeService(session).get_order(user.id, order_id))


@router.post("/orders/{order_id}/pay", response_model=ApiResponse[PaymentResult])
async def pay_order(
    order_id: int, session: DbSession, user: CurrentUser
) -> ApiResponse[PaymentResult]:
    return ApiResponse(
        message="支付成功",
        data=await TradeService(session).pay_order(user.id, order_id),
    )


@router.post("/orders/{order_id}/cancel", response_model=ApiResponse[OrderDetail])
async def cancel_order(
    order_id: int, session: DbSession, user: CurrentUser
) -> ApiResponse[OrderDetail]:
    return ApiResponse(
        message="订单已取消",
        data=await TradeService(session).cancel_order(user.id, order_id),
    )


@router.post("/orders/{order_id}/complete", response_model=ApiResponse[OrderDetail])
async def complete_order(
    order_id: int, session: DbSession, user: CurrentUser
) -> ApiResponse[OrderDetail]:
    return ApiResponse(
        message="已确认收货",
        data=await TradeService(session).complete_order(user.id, order_id),
    )


@router.post(
    "/reviews",
    response_model=ApiResponse[ReviewPublic],
    status_code=status.HTTP_201_CREATED,
)
async def create_review(
    payload: ReviewCreate, session: DbSession, user: CurrentUser
) -> ApiResponse[ReviewPublic]:
    return ApiResponse(
        message="评价发布成功",
        data=await TradeService(session).create_review(user, payload),
    )
